odoo/service/server.py

```diff
@@ -1166,6 +1166,28 @@ def start(preload=None, stop=False):
             # turn on buffering also for wfile, to avoid partial writes (Default buffer = 8k)
             werkzeug.serving.WSGIRequestHandler.wbufsize = -1
     else:
+        if platform.system() == "Linux" and sys.maxsize > 2**32 and 'MALLOC_ARENA_MAX' not in os.environ:
+            # glibc's malloc() uses arenas [1] in order to efficiently handle memory allocation of multi-threaded
+            # applications. This allows better memory allocation handling in case of multiple threads that
+            # would be using malloc() concurrently [2].
+            # Due to the python's GIL, this optimization have no effect on multithreaded python programs.
+            # Unfortunately, a downside of creating one arena per cpu core is the increase of virtual memory
+            # which Odoo is based upon in order to limit the memory usage for threaded workers.
+            # On 32bit systems the default size of an arena is 512K while on 64bit systems it's 64M [3],
+            # hence a threaded worker will quickly reach it's default memory soft limit upon concurrent requests.
+            # We therefore set the maximum arenas allowed to 2 unless the MALLOC_ARENA_MAX env variable is set.
+            # Note: Setting MALLOC_ARENA_MAX=0 allow to explicitely set the default glibs's malloc() behaviour.
+            #
+            # [1] https://sourceware.org/glibc/wiki/MallocInternals#Arenas_and_Heaps
+            # [2] https://www.gnu.org/software/libc/manual/html_node/The-GNU-Allocator.html
+            # [3] https://sourceware.org/git/?p=glibc.git;a=blob;f=malloc/malloc.c;h=00ce48c;hb=0a8262a#l862
+            try:
+                import ctypes
+                libc = ctypes.CDLL('libc.so.6')
+                M_ARENA_MAX = -8
+                assert libc.mallopt(ctypes.c_int(M_ARENA_MAX), ctypes.c_int(2))
```

Comment:
Please note that `assert` statements are dropped when running CPython in optimized mode (`-O` flag) [[source](https://docs.python.org/3.5/reference/simple_stmts.html#the-assert-statement)].

Reply:
The assert statement is already used for similar stuff in other places in Odoo. As long as it's not used for security/safety purpose it's not a problem to use it (here we're talking about a log message)

Reply:
>  (here we're talking about a log message)

The whole `libc.mallopt(ctypes.c_int(M_ARENA_MAX), ctypes.c_int(2))` statement would be stripped in optimized mode, so it essentially would not be executed.

Reply:
@naglis you're quite right, though in practice we don't support running Odoo with CPython's optimized mode (mostly because it's pointless), so I don't think we should make an exception here. Otherwise we'd need to start changing all asserts everywhere, for no good reason, really ;-)